Add I2S driver for LPC devices

mmahadevan108 opened the pull request:
1. Add I2S driver for LPC devices. This PR was tested using the tests inside tests/drivers/i2s. 
2. The LPC DMA driver was also updated to improve DMA efficiency.
3. The tests were updated to account for loopback over separate I2S ports versus the same port. The default setting for the tests should have the original behavior.

NOTE: 
1. I am unable to pass the below test in test_i2s_states.c. 
```
/** @brief Verify all failure cases in RUNNING state.
*
* - Sending START, PREPARE trigger in RUNNING state returns failure.
*/
void test_i2s_state_running_neg(void)

```
I think it is an issue with the test as the test is underflowing resulting in the I2S TX Channel to not be in the RUNNING state.

2. I am seeing an MPU Fault – Data Access Violation when running the tests in user mode, test code below:

```
              /* Now run all tests in user mode */
              ztest_test_suite(i2s_user_loopback_test,
                                           ztest_user_unit_test(test_i2s_tx_transfer_configure_0),
                                           ztest_user_unit_test(test_i2s_rx_transfer_configure_0),
                                           ztest_user_unit_test(test_i2s_transfer_short),
                                           ztest_user_unit_test(test_i2s_transfer_long),
                                           ztest_user_unit_test(test_i2s_rx_sync_start),
                                           ztest_user_unit_test(test_i2s_rx_empty_timeout),
                                           ztest_user_unit_test(test_i2s_transfer_restart),
                                           ztest_user_unit_test(test_i2s_transfer_tx_underrun),
                                           ztest_user_unit_test(test_i2s_transfer_rx_overrun));
              ztest_run_test_suite(i2s_user_loopback_test);

              ztest_test_suite(i2s_user_states_test,
                                           ztest_user_unit_test(test_i2s_tx_transfer_configure_1),
                                           ztest_user_unit_test(test_i2s_rx_transfer_configure_1),
                                           ztest_user_unit_test(test_i2s_state_not_ready_neg),
                                           ztest_user_unit_test(test_i2s_state_ready_neg),
                                           ztest_user_unit_test(test_i2s_state_running_neg),
                                           ztest_user_unit_test(test_i2s_state_stopping_neg),
                                           ztest_user_unit_test(test_i2s_state_error_neg));
              ztest_run_test_suite(i2s_user_states_test);

```

mmahadevan108 committed:
drivers: i2s: Add support for I2S driver for LPC devices

This uses the API's provided by the SDK I2S driver.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: Add support for I2S driver for LPC devices

Add support for I2S driver for LPC devices

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
tests: i2s: Update i2s_api test

1. Provide an option to loopback over separate I2S ports
2. Increase the buffer count in the loopback test to 4
3. Update test_i2s_transfer_rx_overrun test on how to
   handle RX overrun failures

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: dma: Update the LPC DMA driver

1. Update the SDK API's called in the configure implementation.
The DMA_PrepareTransfer and DMA_SubmitTransfer SDK functions
are not recommneded for use. Replaced the call to these SDK
functions with other SDK API's.
2. Fix the implementation the configure function when multiple
blocks are used.
3. Update the dma_reload implementation. The old reload function
would simply abort the transfer. The new implementation reloads
the DMA buffers for transfer.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
tests: i2s: Fix i2s_states_test test suite

1. Update the way test_i2s_state_error_neg test handles
RX overflow.
2. Clean up after test_i2s_state_stopping_neg test in case the
receiving channel is stuck in the STOPPING State if the data
was received prior to calling the STOP trigger.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
tests: i2s: Update i2s_speed test

1. Allow loopback over different I2S ports
2. Add a config option to indicate if the ports
   are separate

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
cc @hakehuang

galak commented on drivers/i2s/Kconfig.mcux_flexcomm:
```diff
@@ -0,0 +1,22 @@
+if I2S_MCUX_FLEXCOMM
+
+config I2S_MCUX_FLEXCOMM_RX_BLOCK_COUNT
+	int "RX queue length"
```
Does it make sense to have these as devicetree properties?

mmahadevan108 replied:
I followed the approach taken by other I2S drivers.

galak commented on soc/arm/nxp_imx/rt6xx/soc.c:
```diff
@@ -159,6 +159,14 @@ static ALWAYS_INLINE void clock_init(void)
+#if (DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(flexcomm1), nxp_lpc_i2s, okay)) && \
+	(DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(flexcomm3), nxp_lpc_i2s, okay))
+	/* attach AUDIO PLL clock to FLEXCOMM1 (I2S1) */
+	CLOCK_AttachClk(kAUDIO_PLL_to_FLEXCOMM1);
```
Might be better to write this as:
```
#if ... flexcomm1...
  CLOCK_AttachClk(kAUDIO_PLL_to_FLEXCOMM1);
#endif
#if ... flexcomm2...
  CLOCK_AttachClk(kAUDIO_PLL_to_FLEXCOMM2);
#endif
...
```

mmahadevan108 replied:
Sure, I will make this change

hakehuang replied:
@mmahadevan108  just reminder this needs changed.

galak reviewed (commented):
Is there a use case for I2S Tx or Rx only?

mmahadevan108 commented on boards/arm/mimxrt685_evk/pinmux.c:
```diff
@@ -278,6 +278,126 @@ static int mimxrt685_evk_pinmux_init(const struct device *dev)
+#endif
+
+	/* Pin is configured as FC3_RXD_SDA_MOSI_DATA */
+	const uint32_t port0_pin23_config = (
```
Good point, I will change this across this file.

mmahadevan108 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -351,10 +367,72 @@ static int dma_mcux_lpc_reload(const struct device *dev, uint32_t channel,
+
+		/* Only one buffer, enable interrupt */
+		xferConfig = DMA_CHANNEL_XFER(0UL, 0UL, 1UL, 0UL,
+									  data->width,
```
Thanks. Done

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,911 @@
+		strm = &dev_data->dma_tx;
+	}
+
+	blk_cfg = &strm->dma_block[0];
```
Thank you. Good catch. I was trying something in the driver, I have changed to remove the 4 block count.

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,911 @@
+
+#define I2S_DMA_CHANNELS(id)				\
+	.dma_tx = {						\
+		.dma_name = UTIL_AND(		\
```
We could have a case of only TX or RX. This checks for this scenario

mmahadevan108 commented on soc/arm/nxp_imx/rt6xx/soc.c:
```diff
@@ -159,6 +159,14 @@ static ALWAYS_INLINE void clock_init(void)
 #endif
 
+#if (DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(flexcomm1), nxp_lpc_i2s, okay)) && \
+	(DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(flexcomm3), nxp_lpc_i2s, okay))
```
This is how it is done for the rest of the drivers, would require a broader change which we can look at outside this PR.

mmahadevan108 commented:
> Is there a use case for I2S Tx or Rx only?

Yes, I think so.

galak commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,908 @@
+	static const struct i2s_mcux_config i2s_mcux_config_##id = {	\
+		.base =							\
+		(I2S_Type *)DT_INST_REG_ADDR(id),			\
+		.clock_name = DT_INST_CLOCKS_LABEL(id),		\
```
Trying to phase out DT_INST_CLOCKS_LABEL, see recent changes to other NXP drivers to use `DEVICE_DT_GET`

mmahadevan108 replied:
Thank you, I will make this change.

MaureenHelm replied:
Thanks for fixing this, but it was done in the wrong commit. Can you move the change into the commit `drivers: i2s: Add support for I2S driver for LPC devices`

mmahadevan108 replied:
Thanks for catching this, I have fixed this.

MaureenHelm commented on boards/arm/mimxrt685_evk/pinmux.c:
```diff
@@ -13,7 +13,7 @@ static int mimxrt685_evk_pinmux_init(const struct device *dev)
 #if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(flexcomm0), nxp_lpc_usart, okay) && CONFIG_SERIAL
 	/* USART0 RX,  TX */
-	const uint32_t port0_pin1_config = (/* Pin is configured as FC0_TXD_SCL_MISO_WS */
+	uint32_t port0_pin1_config = (/* Pin is configured as FC0_TXD_SCL_MISO_WS */
```
Please move the const changes to a separate commit, they appear unrelated to I2S

mmahadevan108 replied:
Sure I will move to a separate commit.

MaureenHelm commented on boards/arm/mimxrt685_evk/pinmux.c:
```diff
@@ -278,6 +278,126 @@ static int mimxrt685_evk_pinmux_init(const struct device *dev)
 	IOPCTL_PinMuxSet(IOPCTL, 1U, 6U, port1_pin6_config);
 #endif
 
+#if (DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(flexcomm1), nxp_lpc_i2s, okay)) && \
+	(DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(flexcomm3), nxp_lpc_i2s, okay)) && \
```
Why use flexcomm1 in the first line and flexcomm3 in the second line? I assume this is a typo?

mmahadevan108 replied:
I have updated this per feedback from @galak.

MaureenHelm replied:
I see what this is doing now, this is needed for the shared signals. That said, can you pull the pinmuxing out of this conditional and use separate conditionals for each instance? The way it's configured now won't work if you only enable one instance.

mmahadevan108 replied:
For the mimxrt685 EVK board we are using both TX and RX I2S. The flexibility to have only one enabled is provided in the common files such as I2S and soc.c

MaureenHelm reviewed (changes requested):
Looks good, just a few nits.

ehughes commented:
> Is there a use case for I2S Tx or Rx only?

Yes,   I2S DMICs.

mmahadevan108 committed:
MXRT600: Add I2S support

Use flexcomm1 and flexcomm3 for I2S support

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
Anything else needed for this PR?

galak commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,902 @@
+/*
+ * Copyright (c) 2020, NXP
```
2021?

mmahadevan108 replied:
Thank you. Done

galak commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,902 @@
+		    sizeof(void *), CONFIG_I2S_MCUX_FLEXCOMM_RX_BLOCK_COUNT);
+
+	/* Get the binding to the DMA device */
+	if (data->dma_tx.dma_name != NULL) {
```
Replace this with DEVICE_DT_GET and DT_DMAS_CTLR_BY_NAME

mmahadevan108 replied:
sure, I will make this change

galak commented on tests/drivers/i2s/i2s_api/Kconfig:
```diff
@@ -0,0 +1,26 @@
+	help
+	  Use separate I2S ports for transmit and receive.
+
+config I2S_TEST_TX_MASTER
```
Do we need to use master/slave terminology here?

galak commented on tests/drivers/i2s/i2s_speed/src/test_i2s_speed.c:
```diff
@@ -10,6 +10,12 @@
 #include <drivers/i2s.h>
 
 #define I2S_DEV_NAME "I2S_0"
```
Seems like these should be renamed to I2S_DEV_NAME_TX, I2S_DEV_NAME_RX

mmahadevan108 replied:
Done, thank you

galak commented on tests/drivers/i2s/i2s_speed/Kconfig:
```diff
@@ -0,0 +1,26 @@
+	help
+	  Use separate I2S ports for transmit and receive.
+
+config I2S_TEST_TX_MASTER
```
Do we need this, is this related to something specific about the NXP hardware for this test?

mmahadevan108 replied:
This is needed as in the loopback test we are using 2 separate I2S ports, one is configured as Master and the other as Slave. The current test is written for a loopback within the same port which is not available in NXP SoC.

galak replied:
Do we have any sense if we did this configuration always regardless if TX & RX are split or not what the implication would be?

Put another way, if on a STM32 or SAM device we set TX_MASTER=y and RX_MASTER=n, what do we think would happen?

mmahadevan108 replied:
This configuration is available for all boards to use. The default behavior is for TX and RX ports to be in slave mode which is how the test currently functions. Anything different from the default can be configured via these configs.

galak replied:
I guess my question was can we also have the settings as you need for NXP and thus remove the Kconfig options.

mmahadevan108 replied:
That would mean conditional code using an NXP specific define. I felt a generic approach would be better suited as individual boards can have their own setting on MASTER and SLAVE ports.

mnkp replied:
The fact that both functions `test_i2s_rx_transfer_configure` and `test_i2s_tx_transfer_configure` configure SPI interface pins as slave is actually a mistake. It works on Atmel SAM E70 platform since we're relying on the internal loopback but that's not a proper setup. We could modify the testcase and configure TX as master and RX as slave, have one fixed configuration.

In principle it would be good to test different configurations: TX master -> RX slave, TX slave -> RX master but we probably shouldn't introduce Kconfig options. Having user select TX master -> RX master can potentially damage hardware since we will have two pins configured as output shorted together.

mmahadevan108 replied:
The idea behind the Kconfig was not for users to try different options. Rather it was a way to have individual developers pick the setting needed for their boards.

mmahadevan108 replied:
Thanks, I have dropped the config options and updated the test code to configure TX as master and RX as slave.

galak reviewed (dismissed):
* Change DMA Devicetree to use new DEVICE_DT_GET and  DT_DMAS_CTLR_BY_NAME
* some questions/minor cleanups related to the test code.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,897 @@
+		break;
+	case I2S_STATE_ERROR:
+		i2s_mcux_rx_stream_disable(dev, true);
+		strm->state = I2S_STATE_READY;
```
Seems wrong. Once we get into the error state the application needs to be notified and undertake relevant steps to recover. This shouldn't happen automatically.

mmahadevan108 replied:
Got it, I will delete these cases and let the application handle it.

mnkp replied:
The implementation in stm32 driver looks good. In case of error the driver disables the stream but preserves the interface state as `I2S_STATE_ERROR`. In your previous implementation you've been changing the interface state to `I2S_STATE_READY`.

The reason why we shouldn't change the interface state to ready is to make sure application is aware that continuity of the data stream was broken. Samples might be missing or corrupted. Driver cannot really recover from this situation. It's application that needs to worry about it.

mmahadevan108 replied:
Sounds good, then I will disable the stream and set state to I2S_ERROR.

mmahadevan108 replied:
Thanks, I have updated the implementation to drop the line switching to `I2S_STATE_READY` state

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,897 @@
+	data->dma_tx.state = I2S_STATE_NOT_READY;
+	data->dma_rx.state = I2S_STATE_NOT_READY;
+
+	LOG_DBG("%s inited", dev->name);
```
Could be changed to
```
	LOG_INF("Device %s initialized", dev->name);
```
It's actually useful for the user to get an indication that the device was initialized.

mmahadevan108 replied:
Sure, I will make this change.

mnkp commented on tests/drivers/i2s/i2s_api/src/test_i2s_states.c:
```diff
@@ -351,6 +351,12 @@ void test_i2s_state_stopping_neg(void)
 
 	ret = rx_block_read(dev_i2s_rx, 0);
 	zassert_equal(ret, TC_PASS, NULL);
+
+	/* This is incase the RX channel is stuck in STOPPING state.
+	 * Clear out the state before running the next test.
+	 */
+	ret = i2s_trigger(dev_i2s_rx, I2S_DIR_RX, I2S_TRIGGER_DROP);
+	zassert_equal(ret, 0, "RX DROP trigger failed");
```
At this point we've sent two data blocks and we read two data blocks. The driver should not be any more in the stopping state. Seems like the driver implementation is the problem.

mmahadevan108 replied:
This in the case where the STOP trigger is called after the data has already been received and the dma_rx callback has finished executing.

mmahadevan108 replied:
I am referring to the STOP trigger that is sent on line 339. I understand the test is verifying return of error conditions and in my case it is passing those parts. However at the end of the test the RX port was left in STOPPING state as the data was already received I feel prior to calling the STOP trigger. Hence I add the lines of code so we can run the next test.

mnkp replied:
OK, I got it. You are right. This needs to be fixed. The testcase relies on the fact that the transmission on the I2S interface is slow. It shouldn't. I propose to fix it differently though. Please remove the DROP trigger that you added and instead move the lines
```
	/* All but one data block read, stop reception */
	ret = i2s_trigger(dev_i2s_rx, I2S_DIR_RX, I2S_TRIGGER_STOP);
	zassert_equal(ret, 0, "RX STOP trigger failed");
```
to just after the following section
```
	/* All data written, flush TX queue and stop the transmission */
	ret = i2s_trigger(dev_i2s_tx, I2S_DIR_TX, I2S_TRIGGER_DRAIN);
	zassert_equal(ret, 0, "TX DRAIN trigger failed");
```
So we will guarantee that the STOP trigger on the RX interface is sent just after DRAIN trigger on the TX interface.

mmahadevan108 replied:
I tried the moving the lines as you suggested. The test now fails in the below lines as the I2S RX port is in the READY state after going to STOPPING and receiving the data packets.

```
	/* Send invalid triggers, expect failure */
	ret = i2s_trigger(dev_i2s_rx, I2S_DIR_RX, I2S_TRIGGER_START);
	zassert_equal(ret, -EIO, NULL);

```

mnkp replied:
Oh, well. It looks like the multiple calls to `i2s_trigger`, `zassert_equal` take too long on your SoC. By the time the testcase arrives at the section testing RX failure triggers the RX is already done with reception.

I propose to put one more data block in the TX queue, so the testcase will have more time to run the checks. I've added one `tx_block_write` and one accompanying `rx_block_read`. I propose to write the testcase as follows.

```
void test_i2s_state_stopping_neg(void)
{
	const struct device *dev_i2s;
	int ret;

	dev_i2s = device_get_binding(I2S_DEV_NAME);
	zassert_not_null(dev_i2s, "device " I2S_DEV_NAME " not found");

	/* Prefill TX queue */
	ret = tx_block_write(dev_i2s, 0, 0);
	zassert_equal(ret, TC_PASS, NULL);
	ret = tx_block_write(dev_i2s, 0, 0);
	zassert_equal(ret, TC_PASS, NULL);

	/* Start reception */
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_START);
	zassert_equal(ret, 0, "RX START trigger failed");

	/* Start transmission */
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_START);
	zassert_equal(ret, 0, "TX START trigger failed");

	ret = tx_block_write(dev_i2s, 0, 0);
	zassert_equal(ret, TC_PASS, NULL);

	ret = rx_block_read(dev_i2s, 0);
	zassert_equal(ret, TC_PASS, NULL);

	/* All data written, flush TX queue and stop the transmission */
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_DRAIN);
	zassert_equal(ret, 0, "TX DRAIN trigger failed");

	/* Send invalid triggers, expect failure */
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_START);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_STOP);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_DRAIN);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_TX, I2S_TRIGGER_PREPARE);
	zassert_equal(ret, -EIO, NULL);

	ret = rx_block_read(dev_i2s, 0);
	zassert_equal(ret, TC_PASS, NULL);

	/* All but one data block read, stop reception */
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_STOP);
	zassert_equal(ret, 0, "RX STOP trigger failed");

	/* Send invalid triggers, expect failure */
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_START);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_STOP);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_DRAIN);
	zassert_equal(ret, -EIO, NULL);
	ret = i2s_trigger(dev_i2s, I2S_DIR_RX, I2S_TRIGGER_PREPARE);
	zassert_equal(ret, -EIO, NULL);

	ret = rx_block_read(dev_i2s, 0);
	zassert_equal(ret, TC_PASS, NULL);
}
```

I have verified that the code works on `sam_e70_xplained` board with `i2s_sam_ssc.c` driver.

mmahadevan108 replied:
This does not work for me. I start seeing I2S RX overflows. Moving the below 2 lines right below the trigger to DRAIN the TX fixes the overflow, however in that case I still see the original issue.
```
	ret = rx_block_read(dev_i2s, 0);
	zassert_equal(ret, TC_PASS, NULL);
```

mmahadevan108 replied:
This the RX FIFO overflowing at the hardware level triggering an interrupt in the driver.

mnkp replied:
But this shouldn't happen. I'm asking to make sure it's not a bug in the driver. Unfortunately, I don't have a MIMXRT685-EVK board to test the PR locally.

mmahadevan108 replied:
@mknp. I have updated the I2S receive implementation to add double buffering. I am not able to pass the test with your changes. However I still see the issue where the RX channel is stuck in STOPPING state. I will add another commit to update the I2S RX DMA implementation.

I will also add a check to not allow transfers below 16-bit till we are able to add a new API.

mnkp replied:
What do you mean by double buffering? I don't think we need to change RX implementation, unless you found a bug.

> I will also add a check to not allow transfers below 16-bit till we are able to add a new API.

Yes, that's probably the best course of action. Let's disable transfers below 16-bits for now. We can add it later when we figure out how to support it in API.

mnkp replied:
After recent modifications to the driver, do the new version of `test_i2s_state_stopping_neg` testcase that I proposed in the previous comment still fail? Is the failure connected with RX FIFO over- or underflow?

mnkp reviewed (changes requested):
Looks good overall. Most of my comments are suggestions, few are real issues.

Could you improve "drivers: dma: Update the LPC DMA driver " commit message. It's not clear what has changed and why.

I haven't finished testcase review yet.

mmahadevan108 commented:
> Looks good overall. Most of my comments are suggestions, few are real issues.
> 
> Could you improve "drivers: dma: Update the LPC DMA driver " commit message. It's not clear what has changed and why.
> 
> I haven't finished testcase review yet.

I have the below in the commit message:
```
    1. Update the SDK API's called in the configure implementation
    2. Update the dma_reload implementation
```

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,902 @@
+
+struct i2s_mcux_data {
+	struct stream dma_rx;
+	struct stream dma_tx;
```
Thanks, I will make this change.

mnkp commented:
> I have the below in the commit message:
> 
> ```
>     1. Update the SDK API's called in the configure implementation
>     2. Update the dma_reload implementation
> ```

Yes, it is not clear why the changes were required and how this new implementation is better than the old one. One would need to review the diff to understand what has changed.

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,897 @@
+{
+	const struct i2s_mcux_config *cfg = dev->config;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *strm = &dev_data->dma_tx;
```
Sure, I will make this change

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,902 @@
+		I2S_TxInit(cfg->base, &fsl_cfg);
+	}
+
+	bytes_per_frame = (uint8_t)((((base->CFG1 & I2S_CFG1_DATALEN_MASK) >>
```
Yes, I will rename the variable name.

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,897 @@
+	if (((base->CFG1 & I2S_CFG1_ONECHANNEL_MASK) == 0U)) {
+		bytes_per_frame = 4U;
+	}
+    /* since DMA do not support 24bit transfer width, use 32bit instead */
```
Thanks, I have fixed this.

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,897 @@
+	}
+
+	bytes_per_frame = (uint8_t)((((base->CFG1 & I2S_CFG1_DATALEN_MASK) >>
+								I2S_CFG1_DATALEN_SHIFT) + 1U) / 8U);
```
Thank you for pointing this. I see this issue, I will update the code.

mmahadevan108 replied:
Code updated

mmahadevan108 commented:
> > I have the below in the commit message:
> > ```
> >     1. Update the SDK API's called in the configure implementation
> >     2. Update the dma_reload implementation
> > ```
> 
> Yes, it is not clear why the changes were required and how this new implementation is better than the old one. One would need to review the diff to understand what has changed.

Got, I have updated the commit message. The I2S tests were failing due to the incorrect DMA implementation.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,899 @@
+						I2S_CFG1_DATALEN_SHIFT) + 1U);
+
+	/* Convert to bytes */
+	if (data_len % 8) {
+		bytes_per_word = (data_len / 8) + 1;
+	} else {
+		bytes_per_word = data_len / 8;
+	}
```
This works, or you could use a little bit simpler implementation as in [get_word_size_bytes()](https://github.com/zephyrproject-rtos/zephyr/blob/c5caa9b916d15212228116a467ee6f5ccdb519ee/drivers/i2s/i2s_sam_ssc.c#L492)

mmahadevan108 replied:
Thanks for pointing this, I have updated the implementation based on your suggestion.

mmahadevan108 commented:
I am unclear why the build is failing.

mnkp reviewed (commented):
Let's make the following changes to the title line of the following commits:

"tests: i2s: Update I2S loopback tests" -> "tests: i2s: Update i2s_api test"
"tests: i2s: Update I2S State test" -> "tests: i2s: Fix i2s_states_test test suite"

And remove from the commit messages mentions of Kconfig option to configure ports as master or slave.

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,901 @@
+	}
+
+	/* Data length in bits */
+	data_len = (uint8_t)(((base->CFG1 & I2S_CFG1_DATALEN_MASK) >>
```
Sure

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,901 @@
+						I2S_CFG1_DATALEN_SHIFT) + 1U);
+
+	/* Convert to bytes */
+	bytes_per_word = (data_len + 7) / 8U;
```
Done

mmahadevan108 commented:
> Let's make the following changes to the title line of the following commits:
> 
> "tests: i2s: Update I2S loopback tests" -> "tests: i2s: Update i2s_api test"
> "tests: i2s: Update I2S State test" -> "tests: i2s: Fix i2s_states_test test suite"
> 
> And remove from the commit messages mentions of Kconfig option to configure ports as master or slave.

Done

MaureenHelm commented:
@galak @mnkp please revisit

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,901 @@
+	/* Convert to bytes */
+	bytes_per_word = (data_len + 7) / 8U;
+
+	/* if one channel is disabled, bytes_per_word should be 4U, user should
+	 * pay attention that when data length is shorter than 16,
+	 * the data format: left data put in 0-15 bit and right data should put in 16-31
+	 */
+	if (((base->CFG1 & I2S_CFG1_ONECHANNEL_MASK) == 0U)) {
+		bytes_per_word = 4U;
+	}
```
Please reference below descriptions

![TransferWidth](https://user-images.githubusercontent.com/16884721/112066307-1b72ca00-8b34-11eb-9cbb-bf56b603bd76.png)

mnkp replied:
When the user is passing data to the `i2s_write` function the memory block is supposed to have all the data tightly packed next to each other. I.e. In case of the standard I2S data format and 8-bit word the frame data should fit in 2 bytes. The next frame data will occupy adjacent 2 bytes.

This is not very well documented, but is mentioned in `struct i2s_config` description.

It looks like this I2S hardware module + DMA engine are not compatible with how the API was designed. They enforce different data layout.

Would you be able to discuss this issue during today's API meeting?

mmahadevan108 replied:
Sure, I plan to attend the call. I am guessing the issue is that the DMA would transferring extra bytes at the end?

mmahadevan108 replied:
I2S IP not friendly to 8 bit data, so in this case, software should convert the tightly pack data to the format that I2S required, that is

![Picture1](https://user-images.githubusercontent.com/16884721/112323522-db206280-8c7f-11eb-85aa-c5983e363578.png)

The 32-24 bits and 15-8 bits are left with empty.

carlescufi commented:
API meeting 2021.03.23:

- No firm conclusion, though options would be:
  - Add a format parameter of sorts to the config function, or a function that returns the required sample size in bytes
  - Treat the read/write functions in this driver as device-specific and so add i2s_lpc_read/write
- Reach out to the SoF folks to try and see how common this occurrence may be

lgirdwood commented:
@mwasko @marc-hb fyi

Generally I2S IP also supports TDM and PCM audio formats. This means the config structure has to be able to cope with more than just I2S and should be capable to describe a TDM hardware data format. e.g. the config should be able to
1) Set rate for MCLK, BCLK and FRAME.
2) set clock providers for MCLK, BCLK and FRAME.
3) Set the frame size, num channels, sample valid bits (as this can be less than (frame size / num channels)).
4) TDM valid slots for tx and rx.
5) timing between FRAME being asserted wrt BCLK and data MSB. e.g. is MSB sent as FRAME is asserted or N BCLKs after. This is usually defined by a mode e.g. DSP_A, DSP_B, I2S_LEFT_JUSTIFIED etc.
6) Some space for flags and quirks specific to IP.

i.e from SOF general DAI (digital audio interface) config with the Intel SSP IPC config (specific to Intel SSP)
```c

/*
 * DAI Configuration.
 *
 * Each different DAI type will have it's own structure and IPC cmd.
 */

#define SOF_DAI_FMT_I2S		1 /**< I2S mode */
#define SOF_DAI_FMT_RIGHT_J	2 /**< Right Justified mode */
#define SOF_DAI_FMT_LEFT_J	3 /**< Left Justified mode */
#define SOF_DAI_FMT_DSP_A	4 /**< L data MSB after FRM LRC */
#define SOF_DAI_FMT_DSP_B	5 /**< L data MSB during FRM LRC */
#define SOF_DAI_FMT_PDM		6 /**< Pulse density modulation */

#define SOF_DAI_FMT_CONT	(1 << 4) /**< continuous clock */
#define SOF_DAI_FMT_GATED	(0 << 4) /**< clock is gated */

#define SOF_DAI_FMT_NB_NF	(0 << 8) /**< normal bit clock + frame */
#define SOF_DAI_FMT_NB_IF	(2 << 8) /**< normal BCLK + inv FRM */
#define SOF_DAI_FMT_IB_NF	(3 << 8) /**< invert BCLK + nor FRM */
#define SOF_DAI_FMT_IB_IF	(4 << 8) /**< invert BCLK + FRM */

#define SOF_DAI_FMT_CBP_CFP	(0 << 12) /**< codec bclk provider & frame provider */
#define SOF_DAI_FMT_CBC_CFP	(2 << 12) /**< codec bclk consumer & frame provider */
#define SOF_DAI_FMT_CBP_CFC	(3 << 12) /**< codec bclk provider & frame consumer */
#define SOF_DAI_FMT_CBC_CFC	(4 << 12) /**< codec bclk consumer & frame consumer */

#define SOF_DAI_FMT_FORMAT_MASK		0x000f
#define SOF_DAI_FMT_CLOCK_MASK		0x00f0
#define SOF_DAI_FMT_INV_MASK		0x0f00
#define SOF_DAI_FMT_CLOCK_PROVIDER_MASK	0xf000

/** \brief Types of DAI */
enum sof_ipc_dai_type {
	SOF_DAI_INTEL_NONE = 0,		/**< None */
	SOF_DAI_INTEL_SSP,		/**< Intel SSP */
	SOF_DAI_INTEL_DMIC,		/**< Intel DMIC */
	SOF_DAI_INTEL_HDA,		/**< Intel HD/A */
	SOF_DAI_INTEL_ALH,		/**< Intel ALH */
	SOF_DAI_IMX_SAI,                /**< i.MX SAI */
	SOF_DAI_IMX_ESAI,               /**< i.MX ESAI */
};

/* general purpose DAI configuration */
struct sof_ipc_dai_config {
	struct sof_ipc_cmd_hdr hdr;
	uint32_t type;		/**< DAI type - enum sof_ipc_dai_type */
	uint32_t dai_index;	/**< index of this type dai */

	/* physical protocol and clocking */
	uint16_t format;	/**< SOF_DAI_FMT_ */
	uint8_t group_id;	/**< group ID, 0 means no group (ABI 3.17) */
	uint8_t reserved8;	/**< alignment */

	/* reserved for future use */
	uint32_t reserved[8];

	/* HW specific data */
	union {
		struct sof_ipc_dai_ssp_params ssp;
		struct sof_ipc_dai_dmic_params dmic;
		struct sof_ipc_dai_hda_params hda;
		struct sof_ipc_dai_alh_params alh;
		struct sof_ipc_dai_esai_params esai;
		struct sof_ipc_dai_sai_params sai;
	};
} __attribute__((packed));

/* SSP Configuration Request - SOF_IPC_DAI_SSP_CONFIG */
struct sof_ipc_dai_ssp_params {
	uint32_t reserved0;
	uint16_t reserved1;
	uint16_t mclk_id;

	uint32_t mclk_rate;	/* mclk frequency in Hz */
	uint32_t fsync_rate;	/* fsync frequency in Hz */
	uint32_t bclk_rate;	/* bclk frequency in Hz */

	/* TDM */
	uint32_t tdm_slots;
	uint32_t rx_slots;
	uint32_t tx_slots;

	/* data */
	uint32_t sample_valid_bits;
	uint16_t tdm_slot_width;
	uint16_t reserved2;	/* alignment */

	/* MCLK */
	uint32_t mclk_direction;

	uint16_t frame_pulse_width;
	uint16_t tdm_per_slot_padding_flag;
	uint32_t clks_control;
	uint32_t quirks;
	uint32_t bclk_delay;	/* guaranteed time (ms) for which BCLK
				 * will be driven, before sending data
				 */
} __attribute__((packed));
```

mnkp commented:
Thanks @lgirdwood for a detailed explanation. The problem here is more specific. When we send audio data to the I2S driver we place it in a large buffer and pass to the driver using one system call. So far the I2S API assumed that the data are always packed in an optimum way. E.g. an 8-bit sample will require one byte of storage, that is to send 16 8-bit samples we will need a buffer of size 16. However, the LPC driver in this PR is not able to cope with such data. Its DMA engine can do 32-bit transfers only. It can transfer one 32-bit sample or two 16-bit samples at minimum.

We were wondering how such hardware limitations are dealt with in case of SOF. Does the application need to query the data layout supported by the driver? You only support one specific data layout?

ALSA library has [snd_pcm_format_physical_width()](https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m___helpers.html#gaa3e0ff7560342e5af2b5c7bd2d63a307) function which lets the application figure out the `sizeof` a single data sample.

lgirdwood commented:
> Thanks @lgirdwood for a detailed explanation. The problem here is more specific. When we send audio data to the I2S driver we place it in a large buffer and pass to the driver using one system call. 

I would just pass the pointer and bytes in this call (we do). The DMA and I2S are configured with other calls first as this allows for continuous playback..

So far the I2S API assumed that the data are always packed in an optimum way. E.g. an 8-bit sample will require one byte of storage, that is to send 16 8-bit samples we will need a buffer of size 16. However, the LPC driver in this PR is not able to cope with such data. Its DMA engine can do 32-bit transfers only. It can transfer one 32-bit sample or two 16-bit samples at minimum.
> 
> We were wondering how such hardware limitations are dealt with in case of SOF. Does the application need to query the data layout supported by the driver? You only support one specific data layout?

The HW link formats are defined in the pipeline topology and programmed by the DAI drivers when streams are in use. This can also mean the pipeline will react to use case input format from a media player and align this format with the I2S configuration (i.e. via SRC or format conversion).

Generally, if an application tries to open a PCM device where the HW format is incompatible with the HW AND no conversion is possible we will fail. 

> 
> ALSA library has [snd_pcm_format_physical_width()](https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m___helpers.html#gaa3e0ff7560342e5af2b5c7bd2d63a307) function which lets the application figure out the `sizeof` a single data sample.

We have something similar to ALSA (disclaimer: I'm also an ALSA developer) in that we ```frame_size``` parameter for the audio PCM data that corresponds to the number of channels * sample size (frame_format) e.g.
```c
/* stream PCM frame format */
enum sof_ipc_frame {
	SOF_IPC_FRAME_S16_LE = 0,
	SOF_IPC_FRAME_S24_4LE,
	SOF_IPC_FRAME_S32_LE,
	SOF_IPC_FRAME_FLOAT,
	/* other formats here */
};
```
i.e. Stereo S24_4LE is 2 * 4 bytes = 8 bytes frame size.  Please note we dont have 8bit formats, this would sound terrible. 16 bit stereo is pretty much the default supported by all codec HW.

This is all that is needed for copying data from memory to memory (i.e. host to DSP). To copy data to peripherals over TDM, PCM or I2S protocols you also need to include a HW frame_size which may be larger that the PCM data frame size (i.e. we have I2S running with 64BCLKs per FRAME for stereo 16bit PCM will mean the link will have 16 unused BCLKs for L and R - HW frame_size is 64 and PCM frame_size is 32..

carlescufi commented:
> I would just pass the pointer and bytes in this call (we do). The DMA and I2S are configured with other calls first as this allows for continuous playback..

@lgirdwood So you mean something like:

```
static inline int i2s_write(const struct device *dev, void *mem_block, size_t size, size_t sample_size)
```
where `sample_size` is either 8-bit or 16-bit?
The main question we are trying to deal with is how to communicate to the driver the format that the data is provided in (eg. for 8-bit samples, whether they all come packed together or each is placed in a 16-bit word in memory), or, alternatively, communicate to the application what format the driver expects. In the former, the driver would fail if the `sample_size` is set to a value not supported by the hardware, in the latter the driver would "trust" that the application has provided the data in the right format.

lgirdwood commented:
> @lgirdwood So you mean something like:
> 
> ```
> static inline int i2s_write(const struct device *dev, void *mem_block, size_t size, size_t sample_size)
> ```
> 
> where `sample_size` is either 8-bit or 16-bit?

Better to configure sample format in another I2S  API call before you call this write(). I would probably use ALSA API for guidance here e.g.

https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m.html#gabc748a500743713eafa960c7d104ca6f

Btw, ALSA has two flavors of write() -  blocking and non blocking. I'm assuming your version will block ? I'm also assuming that your API is not using DMA and directly programming the audio data to the I2S port FIFO ? (otherwise it's better to use a circular audio buffer).

mmahadevan108 committed:
drivers: i2s: Do not support 8-bit word size in LPC I2S driver

The memory block passed by the user to the i2s_write function is
tightly packed next to each other.
However for 8-bit word_size the I2S hardware expects the data
to be in 2bytes which does not match what is passed by the user.
This will be addressed in a separate PR once the zephyr API committee
finalizes on an I2S API for the user to probe hardware variations.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
@mnkp I have rebased and updated this PR to address all pending issues. Should we add a separate issue for the new I2S API.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,965 @@
+	LOG_DBG("dest_burst_length is %d", stream->dma_cfg.dest_burst_length);
+	LOG_DBG("block_count is %d", stream->dma_cfg.block_count);
+
+	dma_config(stream->dev_dma, stream->channel, &stream->dma_cfg);
```
It may be cleaner to put the call to `dma_config` inside `i2s_mcux_config_dma_blocks`. Looking at the function parameters of the two calls

```
i2s_mcux_config_dma_blocks(dev, I2S_DIR_TX, (uint32_t *)buffer);
dma_config(stream->dev_dma, stream->channel, &stream->dma_cfg);
```

It's not clear that one functions prepares data for the other. Also, we would not need to duplicate the numerous `LOG_DBG` entries.

mmahadevan108 replied:
Done.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,965 @@
+		uint8_t num_of_bufs;
+
+		/* Check if we have multiple RX buffers allocated */
+		num_of_bufs = k_mem_slab_num_used_get(stream->cfg.mem_slab);
```
That looks wrong. User can start the stream, stop it and then start again without reading any data from the RX queue. The mere fact that some buffers were taken from the memory pool does not mean they were reserved for the DMA engine. It would be safer to check that `buffer[1]` is not NULL (if we ensure that it is NULL if only one DMA descriptor is used).

mmahadevan108 replied:
This is preparing the DMA for receive. If no data is received and the user issues a STOP then all the DMA buffers will be freed in i2s_purge_stream_buffers.

mmahadevan108 replied:
I have deleted this check as the code allows only receive memory slab with 2 buffers.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,965 @@
+		return ret;
+	}
+
+	if (num_of_bufs > 1) {
+		/*
+		 * Allocate a second receive buffer from SLAB. This will allow
+		 * us to use a Ping-Pong buffer scheme during receive
+		 */
+		ret = k_mem_slab_alloc(stream->cfg.mem_slab, &buffer[1],
+							K_NO_WAIT);
+		if (ret != 0) {
+			LOG_ERR("buffer alloc from mem_slab failed (%d)", ret);
+			return ret;
+		}
+	}
```
If the DMA engine does not work reliably with only 1 descriptor maybe we should enforce that there has to be at least 2 free memory blocks / 2 descriptors to start RX transfer? It seems that if the DMA engine is started with one descriptor then it will always use one descriptor even if more RX blocks becomes available.

mmahadevan108 replied:
Is it a reasonable use case for the user to do RX with a single buffer in the memory slab?

mnkp replied:
No, not really. But if DMA engine holds always 2 RX buffers then the RX buffer pool will have to be a bit larger. Application also needs to 'own' some buffers so it has time to process and free them. There is some cost in requiring 2 free memory blocks / 2 descriptors though reliability is more important. If DMA engine needs 2 buffers to work well we should demand that at least 2 free RX buffers are available when RX transfer is started.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,965 @@
+	case I2S_STATE_STOPPING:
+	case I2S_STATE_RUNNING:
+		/* retrieve buffer from input queue */
+		ret = k_msgq_get(&stream->in_queue, &buffer, K_NO_WAIT);
+		if (ret != 0) {
+			LOG_ERR("get buffer from in_queue %p failed (%d)",
+				&stream->in_queue, ret);
+			break;
+		}
```
If we failed to read data from the queue we need to mark the state of the stream as `I2S_STATE_ERROR`. However, this check could be probably replaced with an ASSERT since it cannot happen that that at this point `stream->in_queue` is empty. That would be a bug in the driver.

mmahadevan108 replied:
Thanks, I have changed to an assert.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,965 @@
+			break;
+		}
+		/* put buffer to output queue */
+		ret = k_msgq_put(&stream->out_queue, &buffer, K_NO_WAIT);
+		if (ret != 0) {
+			LOG_ERR("buffer %p -> out_queue %p err %d", buffer,
+				&stream->out_queue, ret);
+		}
```
If we failed to put the buffer into the `stream->out_queue` we need to mark the state of the stream as I2S_STATE_ERROR and do the clean up. E.g. free the memory slab. Otherwise we have a memory leak here.

mmahadevan108 replied:
Thanks, I have made this change.

mmahadevan108 commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,965 @@
+	 * This will be addressed in a separate PR once the zephyr API committee
+	 * finalizes on an I2S API for the user to probe hardware variations.
+	 */
+	if (i2s_cfg->word_size <= 8) {
+		return -ENOTSUP;
+	}
```
Done

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,940 @@
+	case I2S_STATE_RUNNING:
+		/* retrieve buffer from input queue */
+		ret = k_msgq_get(&stream->in_queue, &buffer, K_NO_WAIT);
+		assert(ret == 0);
```
In Zephyr we should use `__ASSERT_NO_MSG` macro, it's located in `include/sys/__assert.h`.

mmahadevan108 replied:
Thank you, I have changed it.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,940 @@
+	 * Need at least two bffers on the RX memory slab for
+	 * reliable DMA reception.
+	 */
+	if (num_of_bufs == 1) {
+		return -EINVAL;
+	}
```
I think the condition should be `num_of_bufs <= 1`.

mmahadevan108 replied:
Thanks I have changed it.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,940 @@
+		return -EINVAL;
+	}
+
+	ret = k_mem_slab_alloc(stream->cfg.mem_slab, &buffer[0],
+						K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("buffer alloc from mem_slab failed (%d)", ret);
+		return ret;
+	}
+
+	/*
+	 * Allocate a second receive buffer from SLAB. This will allow
+	 * us to use a Ping-Pong buffer scheme during receive
+	 */
+	ret = k_mem_slab_alloc(stream->cfg.mem_slab, &buffer[1],
+						K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("buffer alloc from mem_slab failed (%d)", ret);
+		return ret;
+	}
```
Let's maybe make a small for loop out of this. The code will be easier to read and it might be smaller.

mmahadevan108 replied:
Sounds good.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,940 @@
+
+	i2s_mcux_config_dma_blocks(dev, I2S_DIR_RX, (uint32_t *)buffer);
+
+	/* put buffers in input queue */
+	ret = k_msgq_put(&stream->in_queue, &buffer[0], K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("failed to put buffer in input queue");
+		return ret;
+	}
+
+	if (num_of_bufs > 1) {
+		/* put second buffer in input queue */
+		ret = k_msgq_put(&stream->in_queue, &buffer[1], K_NO_WAIT);
+		if (ret != 0) {
+			LOG_ERR("failed to put buffer in input queue");
+			return ret;
+		}
+	}
```
Also here we could have a for loop, and we can remove `if (num_of_bufs > 1) {` conditional.

mmahadevan108 replied:
Sounds good.

mnkp commented on drivers/i2s/i2s_mcux_flexcomm.c:
```diff
@@ -0,0 +1,940 @@
+			/* allocate new buffer for next audio frame */
+			ret = k_mem_slab_alloc(stream->cfg.mem_slab, &buffer, K_NO_WAIT);
+			if (ret != 0) {
+				if (k_msgq_num_used_get(&stream->in_queue) > 0) {
+					LOG_DBG("wait for data buffer in queue to fill");
+					break;
+				}
```
Now that we enforce that the DMA engine has always 2 descriptors I think we should remove this code.

mmahadevan108 replied:
This is required as alloc could return an error if the slab has only 2 blocks and both blocks are allocated during START. We wait for the second block that is in the input queue to be also filled. We return an error after all blocks have been filled with RX data.

mnkp replied:
This still looks problematic. The DMA engine is set up to work with 2 descriptors. We've taken one block (being pointed to by one of the descriptors) but don't put any replacement. So from this point on the DMA engine is going to work with one descriptor only, until we stop the RX stream.
- I was under impression the DMA engine doesn't work reliably with 1 descriptor only. It needs to have at least 2, so when one of the blocks is filled in it can immediately start filling in another, before the first one is read by this callback.
- We would need to test if the DMA driver works at all if it is configured with 2 DMA descriptors but only 1 is available.

mmahadevan108 replied:
This has been tested with 2 buffers as the i2s_loopback test allocates 2 RX and TX buffers. Unfortunately in the 2 RX Buffer case as you pointed it would NOT be using both slots all the time unless there is a way to queue the buffer as soon as the user frees a memory block.

mnkp replied:
Since we set up DMA engine to work with 2 descriptors it feels like a bad idea to loose 1 descriptor sometime in the middle of the transfer. If `if (ret != 0) {` branch is ever taken from that point on the transfer will continue with only 1 descriptor available to the DMA engine, for the rest of the transfer until user/application sends a request to stop. It's a weakness we probably shouldn't introduce.

It's OK to update the testcases to allocate 3 (or 4) RX buffers. 

In fact, looking at this, it seems like the driver was always running with 1 descriptor only. Application keeps hold of at least 1 buffer, so with 2 buffers total in the pool only 1 is available to the driver. That's probably the reason for the buffer underflow error you've observed in `test_i2s_state_stopping_neg` testcase.

mmahadevan108 replied:
I have updated the implementation to check if can add queue additional buffers in the RX callback. This does not guarantee it will succeed as the user may still be holding on the buffers, however this code provides an opportunity to queue the second buffer.

mnkp replied:
I don't think we should be trying to add to the queue additional buffers in the RX callback, unless I misunderstood what you mean. Let's just make sure we don't lose any. When we start the transfer the DMA engine holds ownership of 2 buffers, if `if (k_msgq_num_used_get(&stream->in_queue) > 0) {` branch is taken it loses ownership of one buffer and will never get it back. Or otherwise: we've just sent one buffer to the application, now we need to get a replacement for it. Let's remove this if branch.

It's OK to force application to prepare enough buffers so that 2 can be always owned by the DMA engine.

Have you tried to run my version of the `test_i2s_state_stopping_neg` testcase when the testcase is changed to allocate 3 or 4 RX buffers in the memory slab?

mmahadevan108 replied:
The code I added increases the number of buffers in the RX queue to 2 if it was short. This may not have an impact in the 2 buffers scenario. 
Yes, I did try the your version and I was able to receive the data without any RX overflows. However the original issue persisted.

mnkp replied:
Now, after removing the `if` branch, the code looks good.

mnkp reviewed (commented):
Thanks, looks good. I still have some comments though.

I propose to rearrange commits a bit. I think we could squash all the following three commits into one:
- soc: Add I2S support for MXRT600
- drivers: i2s: Add support for I2S driver for LPC devices
- drivers: i2s: Update LPC I2S Flexcomm driver for DMA

since the driver will not work without SoC support, and the update to the 2 DMA descriptors for RX is an integral part of the driver.

Also we need to make sure that the commit updating the DMA driver is placed in git history before the commit introducing I2S driver.

mmahadevan108 commented:
> Thanks, looks good. I still have some comments though.
> 
> I propose to rearrange commits a bit. I think we could squash all the following three commits into one:
> 
> * soc: Add I2S support for MXRT600
> * drivers: i2s: Add support for I2S driver for LPC devices
> * drivers: i2s: Update LPC I2S Flexcomm driver for DMA
> 
> since the driver will not work without SoC support, and the update to the 2 DMA descriptors for RX is an integral part of the driver.
> 
> Also we need to make sure that the commit updating the DMA driver is placed in git history before the commit introducing I2S driver.

I have rearranged the commits and squashed the driver folder commits. I would prefer to keep the soc folder commit separate as I have separate commits per folder.

mnkp reviewed (commented):
I think we still need to rearrange commits a bit. When a user checks out the commit "drivers: i2s: Add support for I2S driver for LPC devices " introducing the I2S driver and then adds their own board files the driver should work. That's not the case at the moment since
- soc: Add I2S support for MXRT600
- dts: Add support for I2S driver for LPC devices

are introduced only later in the git history. Also, due to the missing dts commit the driver would likely fail to build. The git history should be bisectable, meaning user should be able to check out any commit and build the project.

mmahadevan108 commented:
@mnkp @galak The build is failing with the below message
```
 DT_INST_FOREACH_STATUS_OKAY(I2S_MCUX_FLEXCOMM_DEVICE)
      |             ^~~~~~~~~~~~~~~~~~~~~
/workdir/zephyr/drivers/i2s/i2s_mcux_flexcomm.c:919:13: error: Macro is deprecated [-Werror]
```
What do you recommend I use in place of DT_INST_FOREACH_STATUS_OKAY

MaureenHelm commented:
> @mnkp @galak The build is failing with the below message
> 
> ```
>  DT_INST_FOREACH_STATUS_OKAY(I2S_MCUX_FLEXCOMM_DEVICE)
>       |             ^~~~~~~~~~~~~~~~~~~~~
> /workdir/zephyr/drivers/i2s/i2s_mcux_flexcomm.c:919:13: error: Macro is deprecated [-Werror]
> ```
> 
> What do you recommend I use in place of DT_INST_FOREACH_STATUS_OKAY

This is probably due to `device_pm_control_nop` now being deprecated. See #34682

mnkp reviewed (commented):
Looks good. The last few things. Let's make sure that
- dts: Add support for I2S driver for LPC devices

commit is placed in the git history before 
- drivers: i2s: Add support for I2S driver for LPC devices

or otherwise when someone checks out the latter commit the driver will not compile.

mmahadevan108 commented on tests/drivers/i2s/i2s_api/Kconfig:
```diff
@@ -0,0 +1,14 @@
+
+source "Kconfig.zephyr"
+
+config I2S_TEST_SEPARATE_LOOPBACK_PORTS
```
I have made this change.

mmahadevan108 commented:
> > @mnkp @galak The build is failing with the below message
> > ```
> >  DT_INST_FOREACH_STATUS_OKAY(I2S_MCUX_FLEXCOMM_DEVICE)
> >       |             ^~~~~~~~~~~~~~~~~~~~~
> > /workdir/zephyr/drivers/i2s/i2s_mcux_flexcomm.c:919:13: error: Macro is deprecated [-Werror]
> > ```
> > 
> > 
> > What do you recommend I use in place of DT_INST_FOREACH_STATUS_OKAY
> 
> This is probably due to `device_pm_control_nop` now being deprecated. See #34682

Thank you.

mmahadevan108 commented:
@mnkp I think I have addressed all comments, can you please check.

mmahadevan108 commented:
@galak. Can this be merged?

MaureenHelm commented:
Thanks @mnkp for the detailed review!
